pandas/tests/test_strings.py

```diff
@@ -34,12 +150,67 @@ def test_api(self):
         assert Series.str is strings.StringMethods
         assert isinstance(Series(['']).str, strings.StringMethods)
 
-        # GH 9184
-        invalid = Series([1])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_dtype(self, box, all_skipna_inferred_dtypes):
+        # one instance of parametrized fixture
+        inferred_type, values = all_skipna_inferred_dtypes
+
+        t = box(values, dtype=object)  # object dtype to avoid casting
+
+        types_passing_constructor = ['string', 'unicode', 'empty',
+                                     'bytes', 'mixed', 'mixed-integer']
+        if inferred_type in types_passing_constructor:
+            # GH 6106, GH 9322
+            assert isinstance(t.str, strings.StringMethods)
+        else:
+            # GH 9184, GH 23011, GH 23163
+            with tm.assert_raises_regex(AttributeError,
+                                        "Can only use .str accessor with "
+                                        "string values!"):
+                t.str
+            assert not hasattr(t, 'str')
+
+    def test_api_mi_raises(self):
+        mi = MultiIndex.from_arrays([['a', 'b', 'c']])
         with tm.assert_raises_regex(AttributeError,
-                                    "only use .str accessor"):
-            invalid.str
-        assert not hasattr(invalid, 'str')
+                                    'Can only use .str accessor with Index, '
+                                    'not MultiIndex'):
+            mi.str
+        assert not hasattr(mi, 'str')
+
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_method(self, box, all_allowed_skipna_inferred_dtypes,
+                            all_string_methods):
+        # one instance per parametrized fixture
+        inferred_type, values = all_allowed_skipna_inferred_dtypes
+        method_name, args = all_string_methods
+
+        t = box(values, dtype=object)  # object dtype to avoid casting
+
+        bytes_allowed = method_name in ['encode', 'decode', 'len']
+        mixed_allowed = method_name not in ['cat']
```

Comment:
The way this test is written/parametrized should make it very transparent which (inferred) dtypes are allowed for which method.